Moves, consolidates, and makes generic [WaitFor]TerminateInstanceTask and friends

ttomsu opened the pull request:
Makes the `TerminateInstances(Stage|Task)` generic. A related stage (`TerminateInstancesAndDecrementServerGroupStage`) has already been made generic, but the tasks were not moved out of the `kato` package into the new `clouddriver` package. 

This PR:
1. Moves the important logic into `TerminatingInstanceSupport`, so that the same server-group-or-search lookup options can be used in both `TerminateInstance` stages and the `WaitForTerminatedInstancesTask`
1. Moves the common tasks into the clouddriver package
1. Deletes the GCE-specific Terminate stages & tasks
1. Fixes https://github.com/spinnaker/spinnaker/issues/660

It also needs a Deck PR (https://github.com/spinnaker/deck/pull/1809) before being merged.

@duftler, @ajordens, @spinnaker/reviewers  - PTAL

anotherchrisberry commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/WaitForTerminatedInstancesTask.groovy:
```diff
@@ -0,0 +1,106 @@
+    def launchTimes = stage.context.launchTimes as List<Long>
+    if (!instanceIds || !launchTimes || instanceIds.size() != launchTimes.size()) {
+      return new DefaultTaskResult(ExecutionStatus.FAILED)
+    }
```
I'd be much happier with this if we changed the data structure to something like:
```json
{
  instances: [
    { id: 'i-12345678', launchTime: 145634534 },
    { id: 'i-32345678', launchTime: 145634533 }
  ]
}
```
Since we're already breaking the existing API by requiring the `launchTimes` field, seems like we're better served by requiring producers to send in data in a format that doesn't require us to merge lists.

ttomsu replied:
Done. This involved a small conversion in the two orca terminate instance stages. Annoyingly, the two respective ops in clouddriver do not have the same APIs, therefor the conversion was slightly different for each. :: shrug ::

cfieber commented:
hmm sorry I just conflicted this a bit with #690

cfieber commented:
the gist of #690 is 
1. poll less frequently
2. if possible do a single lookup of the server group to get the instances of interest rather than a search

it is mostly constrained to the WaitForTerminatedInstancesTask and I think the `allInstancesRebooted` check would benefit from the same decision around doing a per instance lookup only if there is no serverGroupName available

ajordens commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/WaitForTerminatedInstancesTask.groovy:
```diff
@@ -0,0 +1,100 @@
+@Slf4j
+@Component
+class WaitForTerminatedInstancesTask extends AbstractCloudProviderAwareTask implements RetryableTask {
+  long backoffPeriod = 1000
```
I know @cfieber made some equivalent changes ... backoffPeriod should be 10000 (at least).

ttomsu replied:
Working on the merge now...

ajordens commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/TerminateInstancesTask.groovy:
```diff
@@ -45,8 +47,14 @@ class TerminateInstancesTask extends AbstractCloudProviderAwareTask implements T
+  static Map convert(Stage stage) {
+    def operation = [:] << stage.context
+    def instances = operation.remove("instances")
+    operation.instanceIds = instances.collect { it.id }
```
This doesn't look backwards compatible?

ttomsu replied:
The Orca API is changing, but I'm not changing the Clouddriver API. https://github.com/spinnaker/deck/pull/1809/files shows the call to Orca with the instance ID/launch time object. Previously, this stage passed the whole context through, which if you look at what Deck stuffed in there, the key was `instanceIds` with a list of string IDs.

anotherchrisberry replied:
Are we restricted to terminating instances within a single zone in a given operation? Seems like it based on https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/TerminateGoogleInstancesAtomicOperation.groovy#L71

Only asking because I'm trying to get multi-select operations into the UI for instances right now and thinking I need to batch them up by zone...unless you want to change the Clouddriver API? Eh? Don't you want to change the Clouddriver API, too? Eh?

ttomsu replied:
So you're going to make multiple calls if the instances fall within different regions/zones? 

Personally, I'd prefer to keep the Clouddriver operate on a single instance, and make Orca coordinate multiple calls, one for each instance.

cfieber replied:
I don't think we want to restrict to a single instance - in the AWS case the terminate call supports a list of instances and we shouldn't lose the ability to perform that batch operation

I would definitely rather see the orchestration across regions done up in orca than part of the clouddriver API call though

anotherchrisberry commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/WaitForTerminatedInstancesTask.groovy:
```diff
@@ -0,0 +1,100 @@
+    def instanceFromOort
+    return instances.collect { Map instanceFromDeck ->
+      try {
+        instanceFromOort = oortService.getInstance(cloudProvider, stage.context.region as String, instanceFromDeck.id as String)
```
`OortService#getInstance` expects `account` as its first argument, not `cloudProvider` (https://github.com/spinnaker/orca/pull/689/files#diff-6b84a45106ad8e7ecccf4462d6d1c462R69)

ttomsu replied:
Removed use of oortService in favor of oortHelper

ttomsu commented:
I've rebased on top of @cfieber 's commit. It took a bit of finagling, but it works as expected.

Tested by Terminate Instance and Terminate Instance and Decrement SG on both AWS and GCE.

cfieber commented:
mostly LGTM - only comment is around the interaction with `launchTime`

it feels like instead of making the caller supply launchTime as part of the request, the current launchTime could be an output from the terminate instance task

ajordens commented:
@ttomsu - With regards to this being a breaking change... it's not so much `deck` that I'm worried about. We have numerous API users that are sending the old format (aka. `instanceIds`across).

ttomsu commented:
Ok, I updated this PR to move the common logic into a support class. Instance `launchTime`s are now looked up at the Stage execution time and stuffed in the context for later use in the `WaitForTerminatedInstanceTask`. 

I've also updated the Deck PR mentioned in the description. I made a small change, described in https://github.com/spinnaker/deck/pull/1809#discussion_r48562938, but since nothing I saw on the backend used it, I renamed that var to something useful.

PT(Another)L

anotherchrisberry commented:
LGTM - defer to @cfieber on merging

cfieber commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/support/TerminatingInstanceSupport.groovy:
```diff
@@ -0,0 +1,113 @@
+    } else {
+      List<String> instanceIds = stage.context.instanceIds // Used by terminateInstances.
+      if (!instanceIds) {
+        String instanceId = stage.context.instance // Used by terminateInstanceAndDecrementServerGroup.
```
we need to do this ( https://github.com/spinnaker/clouddriver/pull/235 ) in orca at some point too...

cfieber commented on orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/TerminateInstancesStage.groovy:
```diff
@@ -14,32 +14,37 @@
-    def step2 = buildStep(stage, "monitorTermination", MonitorKatoTask)
-    def step3 = buildStep(stage, "waitForTerminatedInstances", WaitForTerminatedInstancesTask)
-    [step1, step2, step3]
+    stage.context[TerminatingInstanceSupport.TERMINATE_REMAINING_INSTANCES] = instanceSupport.remainingInstances(stage)
```
I'm trying to think through why this should happen at StageBuilder time as opposed to as the first operation inside the TerminateInstancesTask with TerminateInstances producing a stageOutput

Feels like that is the more appropriate time to make this lookup given that it should be pulling the launchTime right before the instance is terminated

Also if it was performed inside the Task, then the TerminatingInstance should be possible to be fully populated (instanceId + launchTime) and Immutable?

ttomsu commented:
Moved lookup of launch times to beginning of terminate task and made the domain object immutable as per @cfieber 's request.

Tested successfully against individual instances as well as multiple instances with a snapshot of @anotherchrisberry 's https://github.com/spinnaker/deck/pull/1831 PR

cfieber commented:
LGTM
